### PR TITLE
[cpuid] Update version to 0.7.0 and fix LNK2019

### LINKED ISSUE #41900: [cpuid] update to 0.7.0
### Library name

cpuid

### New version number

0.7.0

### Other information that may be useful (release notes, etc...)

This version of the library introduces support for ARM CPUs. See the [changelog](https://github.com/anrieff/libcpuid/blob/master/ChangeLog#L194-L219) for details.

### PR DESCRIPTION
**Pull request by FrankXie05**
Fix https://github.com/microsoft/vcpkg/issues/41900
Update version to `0.7.0`
Fix `LNK2019`:
```
cpuid_main.c.obj : error LNK2019: unresolved external symbol exec_cpuid referenced in function cpu_exec_cpuid
rdtsc.c.obj : error LNK2019: unresolved external symbol cpu_rdtsc referenced in function cpu_tsc_mark

rdtsc.c.obj : error LNK2019: unresolved external symbol busy_sse_loop referenced in function cpu_clock_by_ic

libcpuid\cpuid.dll : fatal error LNK1120: 3 unresolved externals
```

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [ ] ~~The "supports" clause reflects platforms that may be fixed by this new version.~~
- [ ] ~~Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.~~
- [ ] ~~Any patches that are no longer applied are deleted from the port's directory.~~
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

File: ports/cpuid/fix-LNK2019.patch [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+--- a/libcpuid/CMakeLists.txt
++++ b/libcpuid/CMakeLists.txt
+@@ -21,7 +21,7 @@ if(UNIX)
+   add_compile_definitions(_GNU_SOURCE)
+ endif(UNIX)
+ 
+-if("${MSVC_CXX_ARCHITECTURE_ID}" MATCHES "x64")
++if(MSVC)
+   list(APPEND cpuid_sources masm-x64.asm)
+ endif()
```
Comment (dg0yt) [inline ports/cpuid/fix-LNK2019.patch:12]: Is this really suited for x86?

**Comment by dg0yt** [comment]
And there is arm support for Linux now FreeBSD now.

File: ports/cpuid/fix-LNK2019.patch (hunk 1 above)
Comment (FrankXie05) [reply ports/cpuid/fix-LNK2019.patch:12]: No, from CI, this is only compatible with x64

File: ports/cpuid/vcpkg.json [hunk 2]
```diff
@@ -4,7 +4,7 @@
   "description": "Provides CPU identification for the x86 (and x86_64)",
   "homepage": "https://github.com/anrieff/libcpuid",
   "license": "MIT",
-  "supports": "(x86 | x64 | arm | linux) & !uwp",
+  "supports": "(x86 | x64 | arm) & linux & !uwp",
```
Comment (dg0yt) [inline ports/cpuid/vcpkg.json:7]: IIUC it builds and install on all x86/x64/arm platforms. Only the capabilties vary.
uwp might be a different story.
```suggestion
  "supports": "(arm | x64 | x86) & !uwp",
```

**Comment by FrankXie05** [comment]
Although the error is `C1021`, from the code logic, `libcpuid` does not support `arm64-windows`.

`cpuid_main.c`：
```
#else
# warning This CPU architecture is not supported by libcpuid
	UNUSED(data);
```

```
libcpuid\cpuid_main.c(1455): fatal error C1021: invalid preprocessor command 'warning'
```

**Comment by dg0yt** [comment]
Again, it is expected to build for arm, but act as no-op for unsupported platforms. It is not marked as build-time `#error` but as `#warning`.
The remaining problem is that MSVC doesn't support `#warning`. (It uses `#pragma warning`.) Given that the warning will be buried in build logs of successuful builds, either remove the offending lines, or make an upstreamable patch which adds the MSVC syntax for MSVC.

**Comment by FrankXie05** [comment]
After local testing, it was found that the upstream did not support assembly code files for `arm64`.
```
libcpuid\CMakeFiles\cpuid.dir\masm-x64.asm.obj : fatal error LNK1112: module machine type 'x64' conflicts with target machine type 'ARM64'
```

**Comment by dg0yt** [comment]
... for MSVC windows.

**Comment by FrankXie05** [comment]
> ... for MSVC windows.

Sorry, I don't understand what you mean.

**Comment by dg0yt** [comment]
The arm problem is limited to MSVC. But the manifest doesn't allow any arm.

**Review by Cheney-W — APPROVED** [review]

File: ports/cpuid/fix-MSVC-warning.patch [hunk 3]
```diff
@@ -0,0 +1,13 @@
+diff --git a/libcpuid/cpuid_main.c b/libcpuid/cpuid_main.c
+index ed3db36..b082caa 100644
+--- a/libcpuid/cpuid_main.c
++++ b/libcpuid/cpuid_main.c
+@@ -1452,7 +1452,7 @@ int cpuid_get_raw_data_core(struct cpu_raw_data_t* data, logical_cpu_t logical_c
+ 	}
+ # endif /* PLATFORM_AARCH64 */
+ #else
+-# warning This CPU architecture is not supported by libcpuid
++#pragma message("Warning: This CPU architecture is not supported by libcpuid")
```
Comment (BillyONeal) [inline ports/cpuid/fix-MSVC-warning.patch:10]: I think upstream needs to be OK with this; they presumably blocked it for a reason?

Comment (FrankXie05) [reply ports/cpuid/fix-MSVC-warning.patch:10]: They have agreed to it. https://github.com/anrieff/libcpuid/pull/206

**Review by Cheney-W — DISMISSED** [review]

File: ports/cpuid/vcpkg.json [hunk 4]
```diff
@@ -1,10 +1,10 @@
 {
   "name": "cpuid",
-  "version": "0.6.4",
+  "version": "0.7.0",
   "description": "Provides CPU identification for the x86 (and x86_64)",
   "homepage": "https://github.com/anrieff/libcpuid",
   "license": "MIT",
-  "supports": "(x86 | x64) & !uwp",
+  "supports": "(x86 | x64) & !uwp & !(windows & (arm | arm64))",
```
Comment (dg0yt) [inline ports/cpuid/vcpkg.json:7]:
```suggestion
  "supports": "!uwp & !(windows & arm)",
```

**Review by dg0yt — COMMENTED** [review]
AFAICT the results from the intermediate work were:
- Upstream doesn't test arm64-windows and relies on the author's claim.
- The author didn't manage to successfully build for arm64-windows.
- But arm is supported by upstream except for Windows.

So patches for arm64-windows can be removed from this PR.

**Comment by BillyONeal** [comment]
Thanks for the update and reviews!

**Pull request merged**